### DISCUSSION TITLE
Using Tailwind CSS plugin variants in children?

### DISCUSSION THREAD

**Discussion by peteruithoven** [tailwindlabs/tailwindcss#11526] [Help] [answered]
With the [Tailwind CSS plugin for Headless UI](https://github.com/tailwindlabs/headlessui/tree/main/packages/%40headlessui-tailwindcss) we get variants to easily style Headless UI elements based on whether they are active, for example:
``` jsx
<Menu.Item>
  <a
    className="ui-active:bg-blue-500 ui-active:text-white ui-not-active:bg-white ui-not-active:text-black"
    href="/account-settings"
  >
    Account settings
  </a>
</Menu.Item>
```

But can we do something similar with the children of the Headless UI elements? Below an example where the ui-* work when applied directly to the Listbox.Option but not on it's children. 
``` jsx
<Listbox.Option
  key={key}
  className="flex items-center cursor-default select-none p-2 m-2 rounded-md gap-2 ui-active:bg-primary-15 text-neutral-80 ui-active:text-neutral-90 ui-not-selected:cursor-pointer"
  value={key}
>
  <div className="truncate grow text-sm font-normal ui-selected:font-semibold">
    {label}
  </div>
  <Icon
    Component={Check}
    size={20}
    strokeWidth={2}
    nonScalingStroke
    className="ui-not-selected:invisible text-primary-60"
  />
</Listbox.Option>
```

Maybe something smart with [groups](https://tailwindcss.com/docs/hover-focus-and-other-states#styling-based-on-parent-state) or arbitrary variants? 

Update: I'd love to keep using the clean `ui-selected`, `ui-not-selected` like variants.

A simple example:
https://play.tailwindcss.com/3Yc5WuIhYn

**Comment by wongjn** [comment]
~~Yes, you could use an arbitrary variant like `[[data-headlessui-state~=selected]_&]` or you could use `group` like `group-data-[headlessui-state~=selected]:`, see https://play.tailwindcss.com/kjvJ2ZYviv for examples.~~

Actually unnecessary since `data-headlessui-state` will always exist on `<Listbox.Option>` as per https://github.com/tailwindlabs/tailwindcss/discussions/11526#discussioncomment-6338221

**Reply by peteruithoven** [reply]
I feel like that leaks the abstraction of how the plugin works. I'd prefer to use the `ui-selected`, `ui-not-selected` like variants. But you've inspired me to make a good example using play.tailwindcss.com. I've added it to the original post.

**Reply by wongjn** [reply]
This satisfied your suggestion for something "with [groups](https://tailwindcss.com/docs/hover-focus-and-other-states#styling-based-on-parent-state) or arbitrary variants". It feels unfair to say that this does not meet your expectations and instead move the goalposts after the solution has been presented.

**Reply by peteruithoven** [reply]
You're totally right. And I do appreciate the help. I'm sorry I was unclear about this, I'll clarify the original post.

**Comment by peteruithoven** [comment]
After some experimentation I found that adding something like this to the parent Headless UI element also works. This way I can still use the neater variants from the Tailwind CSS plugin. 
```
[&>svg]:ui-not-selected:invisible
```

Full example:
``` jsx
<Listbox.Option
  key={key}
  className="flex items-center cursor-default select-none p-2 m-2 rounded-md gap-2 ui-active:bg-primary-15 text-neutral-80 ui-active:text-neutral-90 ui-not-selected:cursor-pointer ui-selected:font-semibold [&>svg]:ui-not-selected:invisible"
  value={key}
>
  <div className="truncate grow text-sm">{label}</div>
  <Icon
    Component={Check}
    size={20}
    strokeWidth={2}
    nonScalingStroke
    className=" text-primary-60"
  />
</Listbox.Option>
```
Simple example: https://play.tailwindcss.com/AgMdncTSmq

I would however prefer to add the style classes to the element I want to target.

**Reply by wongjn** [reply]
`ui-not-selected` should work on children: https://play.tailwindcss.com/kMzlACV8v4

**Comment by peteruithoven** [comment]
Being able to add `group` to the headless UI element and then adding something like  `group-ui-not-selected:invisible` to the child would be great.

**Comment by peteruithoven** [comment]
Something weird is going on, the idea of adding `ui-not-selected:invisible` to children of Headless UI elements does just work in a simple example :thinking: 
https://play.tailwindcss.com/zJoZo5YQHD

**Reply by wongjn** [reply]
Consider checking:

- The `className` prop on the `<Icon>` component is applied to the expected element in the rendered HTML.
- The file the code is in is covered by the `content` file globs in the Tailwind configuration.

**Reply by peteruithoven** [reply]
When I copied just the specific generated `li` things seem to work as they should, but when I copy the whole `ul` things stop working. When I removed the  `data-headlessui-state` from the `ul` things work again. I think somehow the `data-headlessui-state="open"`  from the `ul` made the not selected variant still apply? 

Going with the approach of setting a default and then overring it with `ui-selected` does seem to work. 

To clarify. Two examples where I now also include the `data-headlessui-state` in the `ul`:
1. Using `ui-not-selected:invisible`, which doesn't work:
https://play.tailwindcss.com/3Yc5WuIhYn
2. Using  `invisible ui-selected:visible`, which works:
https://play.tailwindcss.com/qKLClPStfS

**Comment by peteruithoven** [comment, marked as answer]
After some experimentation it does actually seem possible to use `ui-selected` like variants, but the `not` doesn't work. Probably because of the `ul` also having a `data-headlessui-state`.


Two examples which include the `data-headlessui-state` in the `ul`:
1. Using ui-not-selected:invisible, which doesn't work:
https://play.tailwindcss.com/3Yc5WuIhYn
2. Using invisible ui-selected:visible, which works:
https://play.tailwindcss.com/qKLClPStfS

Overall simplified example, with both working and not working approach:
https://play.tailwindcss.com/0kVqPTeljj
